pkg/agent/server/modes/host/authenticator.go

```diff
@@ -0,0 +1,127 @@
+// Package host defines authentication and sessions handles when SSH's server is running in host mode.
+package host
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+
+	gliderssh "github.com/gliderlabs/ssh"
+	"github.com/shellhub-io/shellhub/pkg/agent/pkg/osauth"
+	"github.com/shellhub-io/shellhub/pkg/agent/server/modes"
+	"github.com/shellhub-io/shellhub/pkg/api/client"
+	"github.com/shellhub-io/shellhub/pkg/models"
+	log "github.com/sirupsen/logrus"
+	gossh "golang.org/x/crypto/ssh"
+)
+
+// NOTICE: It forces the Authentication struct to implement the Authenticator interface.
+var _ modes.Authenticator = (*Authenticator)(nil)
+
+// Authenticator implements the Authenticator interface when the server is running in host mode.
+type Authenticator struct {
+	// api is a client to communicate with the ShellHub's API.
+	// It is used to get the device public key and validted it against the signature sent by the device.
+	api client.Client
+	// authData is the authentication data received from the API to authenticate the device.
+	authData *models.DeviceAuthResponse
+	// singleUserPassword is the password of the single user.
+	// When it is empty, it means that the single user is disabled.
+	singleUserPassword string
+	// deviceName is the device name.
+	//
+	// NOTICE: It's a pointer because when the server is created, we don't know the device name yet, that is set later.
+	deviceName *string
+}
+
+// NewAuthenticator creates a new instance of Authenticator for the host mode.
+// It receives the api client, the authentication data, the single user password and the device name.
+// The device name is a pointer to a string because when the server is created, we don't know the device name yet, that
+// is set later.
+func NewAuthenticator(api client.Client, authData *models.DeviceAuthResponse, singleUserPassword string, deviceName *string) *Authenticator {
+	return &Authenticator{
+		api:                api,
+		authData:           authData,
+		singleUserPassword: singleUserPassword,
+		deviceName:         deviceName,
+	}
+}
+
+// Password handles the server's SSH password authentication when server is running in host mode.
+func (a *Authenticator) Password(ctx gliderssh.Context, _ string, pass string) bool {
+	log := log.WithFields(log.Fields{
+		"user": ctx.User(),
+	})
+	var ok bool
+
+	if a.singleUserPassword == "" {
+		ok = osauth.AuthUser(ctx.User(), pass)
+	} else {
+		ok = osauth.VerifyPasswordHash(a.singleUserPassword, pass)
+	}
```

Comment:
It would be nice to:

1. use an enum for password authentication type
2. add a debug log informing what is in use

Reply:
I'm working on it in a separated PR.